### PR TITLE
Add Support to Embed .tldr files into .svg files

### PR DESCRIPTION
**Pull request by seflless**
The goal of this PR is to add support for having the underlying file be svg, with the Tldraw state stored as JSON in a comment or some other meta data. This way the file can be used anywhere an image can be used, with the most obvious use case being the README.md of projects or any markdown.

It would work like this:
104091822-444b2900-5280-11eb-8350-1cd9e0579795

Looks like the live reloading of images is [was added to the built-in VS Code Markdown previewer](https://github.com/microsoft/vscode/pull/114083). But it currently isn't working in my testing. So I reached out to the PR's creator to see if I'm doing some wrong or it's broken.

**Todo**

**Future Work**
 - We'll use TldrawApp.copyToSVG() as is for now, which doesn't support including Stickies and Text elements. So we'll be motivated to look into that in another PR. 
 -  - I assume it's embedding fonts that's the pain point for supporting Text in the SVG generation?
 - Ideally Tldraw could have utilities for ensuring custom elements could tap into the toSVG rendering system
 - Doing actual tests and research on making git friendly file formatting would be good. 
   - Research questions: 
     - How do we reduce as much as possible merge conflicts for json or svg+json files? 
       - This unmaintained module was asking similar questions: https://www.npmjs.com/package/gsvg
     - How do we make it as intuitive as possible to hand fix files in their textual format?
     - How hard would it be to make a visual diff/merge tool for .tldr or .tldr.svg files?

**Done**
 - Tried to see if html2canvas worked as is in rendering to PNG, then I assumed I could maybe hack it to render to SVG. No dice and I was making no headway trying to figure out why.

### CODE REVIEW

File: apps/vscode/editor/src/app.tsx [hunk 1]
```diff
@@ -4,14 +4,44 @@ import { Tldraw, TldrawApp, TDFile, TDDocument } from '@tldraw/tldraw'
 import { vscode } from './utils/vscode'
 import { defaultDocument } from './utils/defaultDocument'
 import type { MessageFromExtension, MessageFromWebview } from './types'
+import {toSVG, fromSVG} from "./utils/svgEmbedder";
```
Comment (cagataycali) [inline apps/vscode/editor/src/app.tsx:7]: linter needed here,
```suggestion
import { toSVG, fromSVG } from './utils/svgEmbedder';
```

File: apps/vscode/editor/src/app.tsx [hunk 2]
```diff
@@ -4,14 +4,44 @@ import { Tldraw, TldrawApp, TDFile, TDDocument } from '@tldraw/tldraw'
...
+import type { MessageFromWebview } from '../types'
 
 // Will be placed in global scope by extension
+declare let svgEmbedded:boolean;
+declare let initialFileContent:string;
 declare let currentFile: TDFile
 
+function parseFile(fileContent):TDFile{
+  try {
+    if( svgEmbedded === false ){
```
Comment (cagataycali) [inline apps/vscode/editor/src/app.tsx:18]:
```suggestion
    if (!svgEmbedded) {
```

Comment (seflless) [reply apps/vscode/editor/src/app.tsx:18]: I'm more of a fan of the explicit style one.

File: apps/vscode/editor/src/app.tsx (hunk 1 above)
Comment (steveruizok) [reply apps/vscode/editor/src/app.tsx:7]: I think our auto linter isn't running on push. Thanks for the eyes though!

File: apps/vscode/editor/src/app.tsx (hunk 2 above)
Comment (cagataycali) [reply apps/vscode/editor/src/app.tsx:18]: I'm also not a fan of any styling 🗡️ 
But what if we have a standard format for everybody already?

https://github.com/tldraw/tldraw/blob/main/.prettierrc

Comment (cagataycali) [reply apps/vscode/editor/src/app.tsx:18]: `svgEmbeded` is `boolean` so, we're not expecting to have a falsy value.
That's means to us, you're already checking your condition explicitly as your definition.

**Comment by steveruizok** [comment]
Just checking the status on this one: is it still something we want to ship? What would still need to be done to get it over the line?

**Comment by seflless** [comment]
One issue was that SVG exporting doesn't work well yet fully for text. Example.

### Tldraw
Screen Shot 2022-01-30 at 4 48 56 PM


### Exported SVG
Screen Shot 2022-01-30 at 4 50 37 PM

I think we also wanted to think more about what the SVG should be, as there are multiple pages and potentially huge diagrams that will just not make sense thumb-nailed as they'll be small.

So if someone wants to take on getting text rendering tightened up, then that might move this along. I'm a little heads down on other stuff, or I'd take it on right now. I'll get to it eventually if someone else doesn't.

**Comment by seflless** [comment]
Test file for reference.

[svg-tester.tldr.zip](https://github.com/tldraw/tldraw/files/7967007/svg-tester.tldr.zip)

**Comment by seflless** [comment]
You know what, I'm pretty passionate about this one. I'll carve out some time to look at the SVG rendering code and see if I can apply some of my pretty fresh knowledge of extracting DOM layouts per character.

**Comment by seflless** [comment]
Do you already have any more elaborate .tldr files design to test SVG exporting?

**Comment by steveruizok** [comment]
Closing this one, though I'd still like to see it. We've got a slightly better way of handling svg strings now, so we may be able to compress and include in both SVGs and PNGs. My concern would be the effect on file size, but there could be a UI to opt-in.

**Pull request closed without merging**